### ISSUE TITLE
Testing: Add TypeScript types validation for modules

### ISSUE THREAD

**Issue by aduth** [WordPress/gutenberg#18838]
Previously: #17014

While Gutenberg packages are not authored in TypeScript, we can still benefit from its [JavaScript type checking](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html) using the JSDoc we already write. This will bring us some benefit of type safety, even as we continue to write modules with JavaScript, not TypeScript.

**Progress:**

All of our modules are to be opted-in to type checking. The progress thus far:

- [x] a11y (#19096)
- [ ] annotations
- [x] api-fetch
    - #29685 
    - https://github.com/WordPress/gutenberg/pull/29719
    - https://github.com/WordPress/gutenberg/pull/30161
- [x] autop (#20669)
- [ ] babel-plugin-import-jsx-pragma
- [ ] babel-plugin-makepot
- [ ] babel-preset-default
- [ ] base-styles
- [x] blob (#19092)
- [ ] block-directory
- [ ] block-editor (#21362, ...)
- [ ] block-library
- [ ] block-serialization-default-parser
- [ ] block-serialization-spec-parser
- [ ] blocks
- [ ] browserslist-config
- [ ] components
     - https://github.com/WordPress/gutenberg/pull/26066
     - https://github.com/WordPress/gutenberg/pull/26078
     - https://github.com/WordPress/gutenberg/pull/29792
- [x] compose (~#21580~ https://github.com/WordPress/gutenberg/pull/27344)
- [ ] core-data
     - https://github.com/WordPress/gutenberg/pull/38666
     - https://github.com/WordPress/gutenberg/pull/39025 (merged in small chunks!)
     - https://github.com/WordPress/gutenberg/pull/41594
- [ ] create-block
- [ ] custom-templated-path-webpack-plugin
- [ ] data
     - https://github.com/WordPress/gutenberg/pull/39081
     - https://github.com/WordPress/gutenberg/pull/41596
- [ ] data-controls
- [x] date https://github.com/WordPress/gutenberg/pull/29789
- [x] dependency-extraction-webpack-plugin (#22498)
- [x] deprecated (https://github.com/WordPress/gutenberg/pull/26429)
- [ ] docgen
- [x] dom (split into multiple smaller PRs)
    - https://github.com/WordPress/gutenberg/pull/29682
    - https://github.com/WordPress/gutenberg/pull/29787
    - https://github.com/WordPress/gutenberg/pull/30412
    - https://github.com/WordPress/gutenberg/pull/30386
- [x] dom-ready (#19089)
- [ ] e2e-test-utils
- [ ] e2e-tests
- [ ] edit-post
- [ ] edit-site
- [ ] edit-widgets
- [ ] editor
- element
  - [x] first pass (#21248)
  - [x] needs work after #21613 
  - [ ] use stricter types config
- [ ] env
- [x] escape-html (#20669)
- [ ] eslint-plugin
- [ ] format-library
- [x] hooks (https://github.com/WordPress/gutenberg/pull/26430)
- [x] html-entities (#20669)
- [x] i18n (#19099)
- [ ] interface
- [x] icons (#21487, #21613, #21781)
- [x] is-shallow-equal (#19090)
- [ ] jest-console
- [ ] jest-preset-default
- [ ] jest-puppeteer-axe
- [ ] keyboard-shortcuts
- [x] keycodes (#19520)
- [ ] library-export-default-webpack-plugin
- [ ] list-reusable-blocks
- [ ] media-utils
- [ ] notices
- [ ] npm-package-json-lint-config
- [ ] nux
- [ ] plugins
- [ ] postcss-themes
- [x] prettier-config (#21053)
- [x] primitives (#21482, #21613, #21781)
- [x] priority-queue (#18997)
- [x] project-management-automation (#20850)
- [ ] redux-routine (#21313)
- [ ] rich-text
- [ ] scripts
- [ ] server-side-render
- [ ] shortcode
- [x] token-list (#18839)
- [x] url (#17014)
- [ ] viewport
- [x] warning (#19317)
- [x] wordcount (#22077)

**Guidelines:**

See [the packages README](https://github.com/WordPress/gutenberg/blob/master/packages/README.md#typescript) for details on TypeScript usage. In summary, a package can opt-in by following these steps:

- Add a `tsconfig.json` to the package root. The [`@wordpress/i18n` package](https://github.com/WordPress/gutenberg/blob/c5fab6d23ff7478096942412bc20dc30654ca82e/packages/i18n/tsconfig.json) has a good example of a basic config.
- Add `"types": "build-types"` to the package's `package.json` so that consumers will pick up the published types from the package.
- Add the package to the root [`tsconfig.json` `references`](https://github.com/WordPress/gutenberg/blob/master/tsconfig.json).

For single files, you can also include a `// @ts-check` line at the top of the file. You should use this line when temporarily testing the impact of type checking, or if you are working to convert a very large package one file at a time.

The `build:package-types` npm script will be run as part of precommit, which handles type checking. It can be very helpful to watch the types build and work through issues surfaced by the compiler as you add types. To watch the types build, run:

```sh
npm run build:package-types -- --watch
```

You may also need to improve existing JSDoc, as it was written based on assumed usage. Type checking will enforce its validity, and in many cases our JSDoc is invalid.

For syntax and usage, reference:
  - https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#javascript-documentation-using-jsdoc
  - https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/
  - https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#supported-jsdoc
  - https://jsdoc.app/

**Comment by aduth** [comment]
cc @dsifford

**Comment by dsifford** [comment]
> It should be okay to use DefinitelyTyped (@types/*) to reference types of external packages

This is good guidance for now while this is being worked on, but once the repo gets all of the JSDoc finished up I would suggest referencing your internal JSDoc definitions too keep things consistent. This can be done using `tsconfig.json`'s `paths` configuration. External types can still and should still be referenced where appropriate.

So for example...

The `tsconfig.json` would be modified with the following new values...

```json
{
  "compilerOptions": {
    "baseUrl": "./packages",
    "paths": {
      "@wordpress/*": [
        "*"
      ]
    }
  }
}
```

And then you'd be able to import types the same as before...

```ts
// Assume this is in a file from some other package that isn't @wordpress/element
/**
 * @typedef {import('@wordpress/element').ComponentType} ComponentType
 */ 
```

(this is off the top of my head, I may be slightly incorrect in the exact implementation, but you should get the idea).

**Comment by aduth** [comment]
@dsifford Yes, definitely, I think for WordPress dependencies we should aim for what you propose.

There was some previous related discussion to this:

- Consideration for what should be part of the "public types API" of our modules: https://github.com/WordPress/gutenberg/pull/18808#discussion_r352255691
- Using TypeScript 3.7 new `--declaration` parameter for emitting first party types from JSDoc
   - https://github.com/WordPress/gutenberg/issues/15027#issuecomment-559271198
   - https://andrewduthie.com/2019/11/27/typescript-tooling-for-your-javascript-projects/#comment-259

**Comment by dsifford** [comment]
Yep, that `--declaration` flag is a pretty big game-changer. Though, the types still won't be quite as good as those written in typescript due to limited support for generics and overloads.

**Comment by sirreal** [comment]
I've started exploring generated `.d.ts` declarations in #18942.

As part of that PR, I typed `i18n` and `is-shallow-equal`. Shall I extract the changes to those modules so they can be landed?

**Comment by aduth** [comment]
> As part of that PR, I typed `i18n` and `is-shallow-equal`. Shall I extract the changes to those modules so they can be landed?

If you're willing to, I think that would make review of the distinct parts much easier, yes.

**Comment by DanielRosenwasser** [comment]
This seems like a potentially good opportunity for our team (TypeScript) to get some feedback! Let us know if you'd like to chat at some point, or just have any feedback on adding `checkJs` validation.

Also, CC @weswigham and @sandersn in case you're interested in a fairly large JSDoc'd codebase!

**Comment by aduth** [comment]
@DanielRosenwasser Thanks for offering your assistance and for being receptive to the feedback!

Speaking for myself in having implemented JSDoc-based types enforcement for a few of the packages in this project (and [written about](https://andrewduthie.com/2019/11/27/typescript-tooling-for-your-javascript-projects/) and used this approach in a few of my own [personal](https://github.com/aduth/streamlens) [projects](https://github.com/aduth/tannin)), I've generally been pleasantly surprised with how well it "just works".

The struggles we've encountered thus far are usually not a fault of TypeScript itself, but moreso on interoperability with other tooling. For example, we use [`eslint-plugin-jsdoc`](https://www.npmjs.com/package/eslint-plugin-jsdoc) to validate our JSDoc, but because this doesn't [yet](https://github.com/jsdoctypeparser/jsdoctypeparser/issues/50) have complete support for TypeScript types syntaxes, it often leads to a number of false positives. Similarly, we've implemented our own [automated documentation tool](https://github.com/WordPress/gutenberg/tree/master/packages/docgen) which uses the (now-defunct) [Doctrine parser](https://github.com/eslint/doctrine), which also fails to understand this syntax (see also #19571, #18045). In general, I think this speaks more to the insufficiencies of the base JSDoc standard, which in turn have motivated various syntax extensions (TypeScript, [Closure Compiler](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler)), ultimately resulting in a mixed landscape of support for "JSDoc" as somewhat ambiguous in its interpreted standard.

There's been some challenge in learning and applying these syntax extensions, and finding a good balance of just how far we want to take it. I'm sure this is a common theme in any project which implements TypeScript (especially migrating from JavaScript), but I often find it more acute in using the TypeScript syntaxes in JSDoc, where it has a greater impact on line-length and readability (trying to fit the type definition within a single line `@param` type, for example). I suspect these are growing pains which will evolve over time, including with improved guidance around applying these conventions (e.g. #18920).

In case there's interest in the history surrounding this project's relationship with TypeScript, it's been a topic of discussion in several of WordPress' weekly JavaScript meetings:

- https://make.wordpress.org/core/2019/12/04/javascript-chat-summary-december-3-2019/
- https://make.wordpress.org/core/2019/12/02/javascript-chat-summary-november-26-2019/
- https://make.wordpress.org/core/2019/08/01/javascript-chat-summary-july-30-2019/
- https://make.wordpress.org/core/2019/04/17/javascript-chat-summary-april-16th-2019/
- https://make.wordpress.org/core/2019/03/13/javascript-chat-summary-march-12th-2019/

**Comment by sirreal** [comment]
I've updated the description according to current setup now that #18942 has landed. I'm unsure whether this note still applies, but I've left it:

> For single files, you can also include a `// @ts-check` line at the top of the file. You should use this line when temporarily testing the impact of type checking, or if you are working to convert a very large package one file at a time.

**Comment by annezazu** [comment]
Closing this out as the work has either been completed or is no longer relevant. Happy to reopen if we need to but, at this point, a new issue is likely best.
